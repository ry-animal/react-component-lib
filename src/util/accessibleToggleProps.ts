export function accessibleToggleProps(
  label: string,
  controls: string,
  expanded: boolean,
  popupType?: "menu"
) {
  return {
    "aria-controls": controls,
    "aria-expanded": expanded,
    "aria-haspopup": popupType || true,
    "aria-label": label,
    "aria-pressed": expanded,
    role: "button",
  };
}
