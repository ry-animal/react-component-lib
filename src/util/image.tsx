import { ISelectedFile } from "src/components/Forms/FileUpload/ISelectedFile";

export const makeDataUrl = (selectedFile: ISelectedFile) =>
  URL.createObjectURL(selectedFile.file);
