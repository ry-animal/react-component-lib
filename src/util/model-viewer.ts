/*
 * This file is a hacky way of allowing any other component to
 * use the <model-viewer /> tag just by doing:
 *   import "../../../util/model-viewer";
 *
 * The authors of the package do not expose it in a React-friendly
 * way, see: https://github.com/google/model-viewer/issues/1502
 *
 * The component documentation is at https://modelviewer.dev/docs/
 */

import "@google/model-viewer";

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    export interface IntrinsicElements {
      "model-viewer": {
        src?: string;
        alt?: string;
        poster?: string;
        autoplay?: boolean;
        exposure?: number;
        "camera-controls"?: boolean;
        "environment-image"?: string;
        "seamless-poster"?: string;
        loading?: "auto" | "lazy" | "eager";
        reveal?: "auto" | "interaction" | "manual";
      } & React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLElement>,
        HTMLElement
      >;
    }
  }
}
