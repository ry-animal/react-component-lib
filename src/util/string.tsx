const DEFAULT_LOCALE = "en-US";

export const truncateEthAddress = (str: string): string => {
  if (str.length) {
    const numStartingCharacters = 6;
    const numTrailingCharacters = 4;
    return `${str.substring(0, numStartingCharacters)}...${str.substring(
      str.length - numTrailingCharacters,
      str.length
    )}`;
  }
  return "";
};

export const formatUSDPriceString = (value: number): string => {
  const minimumFractionDigits = 2;
  const maximumFractionDigits = 2;
  return `$${value.toLocaleString(DEFAULT_LOCALE, {
    minimumFractionDigits,
    maximumFractionDigits,
  })}`;
};

export const formatETHPriceString = (priceInEth: number): string => {
  const minimumFractionDigits = 0;
  const maximumFractionDigits = 8;
  const maxDigitsWithDecimal = 8;

  let formatted = `${priceInEth
    .toLocaleString(DEFAULT_LOCALE, {
      minimumFractionDigits,
      maximumFractionDigits,
    })
    .replace(/,/gu, "")
    .substring(0, maxDigitsWithDecimal)}`;

  if (formatted.endsWith(".")) {
    formatted = formatted.slice(0, -1);
  }
  return formatted;
};

const evaluateNumberLessThanOne = (num: number) => {
  if (num === 0) {
    return "0";
  }

  const minDecimalValue = 0.0001;
  if (num < minDecimalValue) {
    return minDecimalValue.toString();
  }
  // To round decimal to 4 digits
  const maxDigits = 10000;
  return `${Math.round(num * maxDigits) / maxDigits}`;
};

// Abbreviate Number https://stackoverflow.com/a/9462382

export const abbreviateNumber = (num: number): string => {
  const lookup = [
    { value: 1, symbol: "" },
    { value: 1e3, symbol: "k" },
    { value: 1e6, symbol: "M" },
    { value: 1e9, symbol: "B" },
    { value: 1e12, symbol: "T" },
    { value: 1e15, symbol: "P" },
    { value: 1e18, symbol: "E" },
  ];

  const rx = /\.0+$|(?<id>\.[0-9]*[1-9])0+$/u;
  const digits = 1;

  const item = lookup
    .slice()
    .reverse()
    .find((x) => num >= x.value);
  return item
    ? (num / item.value).toFixed(digits).replace(rx, "$1") + item.symbol
    : evaluateNumberLessThanOne(num);
};

export const getDomainFromRegex = (url: string) => {
  const matches = url.match(/:\/\/(?<id>.[^/]+)/u);
  // Extract domain (will be null if no match is found)
  return matches && matches[1];
};
