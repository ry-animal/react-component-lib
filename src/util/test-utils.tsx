import { render, RenderOptions } from "@testing-library/react";
import "jest-styled-components";
import * as React from "react";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { LIGHT_THEME } from "../theme";

interface ProvidersProps {
  children: React.ReactNode;
}

function Providers({ children }: ProvidersProps) {
  return (
    <BrowserRouter>
      <ThemeProvider theme={LIGHT_THEME}>{children}</ThemeProvider>
    </BrowserRouter>
  );
}

const customRender = (
  ui: JSX.Element,
  options?:
    | RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement
      >
    | undefined
) => render(ui, { wrapper: Providers, ...options });

// Duplicated instead of import "../../__mocks__/matchMediaMock"; because jest has trouble resolving imports
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// Re-export everything
export * from "@testing-library/react";
// Override render method
export { customRender as render };
