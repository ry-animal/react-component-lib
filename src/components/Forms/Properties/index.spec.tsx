import userEvent from "@testing-library/user-event";
import { Properties } from ".";
import { render, screen } from "../../../util/test-utils";

describe("<Properties />", () => {
  const sampleProps = () => ({
    onChange: jest.fn(),
    title: "Properties",
  });

  describe("rendering", () => {
    test("should render properties and matches snapshot", () => {
      expect(
        render(<Properties {...sampleProps()} />).asFragment()
      ).toMatchSnapshot();
    });

    test("should render properties and matches snapshot (with title)", () => {
      expect(
        render(
          <Properties {...sampleProps()} title="Custom title" />
        ).asFragment()
      ).toMatchSnapshot();
    });

    test("should render properties and matches snapshot (with subtitle error)", () => {
      render(
        <Properties {...sampleProps()} subTitleError="Custom subtitle error" />
      );

      expect(screen.queryAllByText("Custom subtitle error").length).toBe(1);
    });

    test("should render properties and matches snapshot (with placeholders)", () => {
      const keyPlaceholder = "e.g. color";
      const valuePlaceholder = "e.g. blue";
      expect(
        render(
          <Properties
            {...sampleProps()}
            keyPlaceholder={keyPlaceholder}
            valuePlaceholder={valuePlaceholder}
          />
        ).asFragment()
      ).toMatchSnapshot();
    });

    test("should render properties (with initial values and error alerts)", () => {
      render(
        <Properties
          {...sampleProps()}
          properties={[
            {
              errorLabelMessage: "",
              errorValueMessage:
                "Our platform does not allow inappropriate language.",
              key: "hello",
              value: "inappropriate message",
            },
          ]}
        />
      );

      const expectedNumberOfInputs = 2;
      const textboxFields = screen.queryAllByRole("textbox");
      expect(textboxFields.length).toBe(expectedNumberOfInputs);
      expect(textboxFields[0].id).toStrictEqual("hello");
      expect(textboxFields[1].id).toStrictEqual("inappropriate message");

      const expectedNumberOfErrorAlerts = 1;
      const alerts = screen.queryAllByRole("alert");
      expect(alerts.length).toBe(expectedNumberOfErrorAlerts);
      expect(alerts[0].innerHTML).toStrictEqual(
        "Our platform does not allow inappropriate language."
      );
    });
  });

  describe("event handling", () => {
    test("onChange is called", () => {
      const props = sampleProps();
      render(<Properties {...props} />);
      const [input] = screen.getAllByRole("textbox");
      userEvent.type(input, "some text");
      expect(props.onChange).toHaveBeenCalled();
    });

    test("editing input updates state", async () => {
      const props = sampleProps();
      render(<Properties {...props} />);
      const [inputKey, inputValue] = screen.getAllByRole(
        "textbox"
      ) as HTMLInputElement[];
      const propertyLabelText = "property label";
      userEvent.type(inputKey, propertyLabelText);
      expect(inputKey.value).toBe(propertyLabelText);

      const propertyValueText = "property value";
      userEvent.type(inputValue, propertyValueText);
      expect(inputValue.value).toBe(propertyValueText);
    });

    test("add button creates another input row", () => {
      render(<Properties {...sampleProps()} />);
      const addButton = screen.getByRole("button", { name: "Add Row" });
      userEvent.click(addButton);
      const expectedNumberOfInputs = 4;
      expect(screen.getAllByRole("textbox").length).toBe(
        expectedNumberOfInputs
      );
    });

    test("remove button removes input row", () => {
      render(<Properties {...sampleProps()} />);
      const removeButton = screen.getByRole("button", {
        name: "Remove property",
      });
      userEvent.click(removeButton);
      const expectedNumberOfInputs = 0;
      expect(screen.queryAllByRole("textbox").length).toBe(
        expectedNumberOfInputs
      );
    });
  });
});
