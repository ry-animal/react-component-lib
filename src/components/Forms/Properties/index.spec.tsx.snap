// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Properties /> rendering should render properties and matches snapshot (with placeholders) 1`] = `
<DocumentFragment>
  .c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex: unset;
  -ms-flex: unset;
  flex: unset;
  line-height: 1;
  gap: 8px;
}

.c8 {
  text-align: center;
  border: none;
  outline: none;
  overflow: hidden;
  cursor: pointer;
  pointer-events: auto;
  -webkit-transition: background-color 0.25s ease-out;
  transition: background-color 0.25s ease-out;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  min-height: 40px;
  font-weight: 600;
  padding: 0 28px;
  font-size: 16px;
  border-radius: 4px;
  background-color: transparent;
  padding: 2px 1px 2px 1px;
  min-height: 16px;
  border-radius: 0px;
  color: #151515;
}

.c8:disabled {
  opacity: 20%;
  cursor: default;
  pointer-events: none;
}

.c8:hover:not([disabled]),
.c8:focus-visible:not([disabled]) {
  color: #404040;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-family: 'Poppins',sans-serif;
  position: relative;
}

.c4 {
  color: #151515;
  font-family: 'Poppins',sans-serif;
  font-weight: 600;
  font-size: 16px;
  line-height: 20px;
  margin-bottom: 8px;
}

.c6 {
  color: #151515;
  background-color: #F9F9F9;
  width: 100%;
  max-width: 100%;
  min-width: 100%;
  font-size: 18px;
  font-family: 'Poppins',sans-serif;
  font-weight: 600;
  line-height: 22px;
  border-radius: 4px;
  border: 1px solid #151515;
  padding: 8px 16px;
  box-sizing: border-box;
  height: 48px;
}

.c6:disabled {
  background: #EEEEEE;
  border: 1px solid #A4A4A4;
}

.c6:invalid {
  border: 2px solid #BC2828;
}

.c6::-webkit-input-placeholder {
  color: #A4A4A4;
  font-size: 18px;
  font-weight: 400;
}

.c6::-moz-placeholder {
  color: #A4A4A4;
  font-size: 18px;
  font-weight: 400;
}

.c6:-ms-input-placeholder {
  color: #A4A4A4;
  font-size: 18px;
  font-weight: 400;
}

.c6::placeholder {
  color: #A4A4A4;
  font-size: 18px;
  font-weight: 400;
}

.c6:focus {
  outline: 1px solid #151515;
}

.c5 {
  position: relative;
}

.c0 {
  color: #151515;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 12px;
  color: #151515;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  width: 100%;
  gap: 12px;
  color: #151515;
}

.c7 {
  color: #151515;
  background: transparent;
  border: none;
  cursor: pointer;
  height: 48px;
  right: 0;
  top: 32px;
  position: relative;
  margin-left: -16px;
  -webkit-transform: scale(0.8);
  -ms-transform: scale(0.8);
  transform: scale(0.8);
}

<div>
    <p
      class="c0"
    >
      Properties
    </p>
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for=""
          >
            Label
          </label>
          <div
            class="c5"
          >
            <input
              aria-invalid="false"
              aria-label="Label"
              class="c6"
              id=""
              name=""
              placeholder="e.g. color"
              value=""
            />
          </div>
        </div>
        <div
          class="c3"
        >
          <label
            class="c4"
            for=""
          >
            Value
          </label>
          <div
            class="c5"
          >
            <input
              aria-invalid="false"
              aria-label="Value"
              class="c6"
              id=""
              name=""
              placeholder="e.g. blue"
              value=""
            />
          </div>
        </div>
        <button
          aria-label="Remove property"
          class="c7"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6L6 18"
              stroke="#151515"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M6 6L18 18"
              stroke="#151515"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </div>
    <button
      aria-label="Add Row"
      class="c8"
    >
      <span
        class="c9"
      >
        <img
          alt="Add Row icon"
          src="[object Object]"
        />
        Add Row
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`<Properties /> rendering should render properties and matches snapshot (with title) 1`] = `
<DocumentFragment>
  .c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex: unset;
  -ms-flex: unset;
  flex: unset;
  line-height: 1;
  gap: 8px;
}

.c8 {
  text-align: center;
  border: none;
  outline: none;
  overflow: hidden;
  cursor: pointer;
  pointer-events: auto;
  -webkit-transition: background-color 0.25s ease-out;
  transition: background-color 0.25s ease-out;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  min-height: 40px;
  font-weight: 600;
  padding: 0 28px;
  font-size: 16px;
  border-radius: 4px;
  background-color: transparent;
  padding: 2px 1px 2px 1px;
  min-height: 16px;
  border-radius: 0px;
  color: #151515;
}

.c8:disabled {
  opacity: 20%;
  cursor: default;
  pointer-events: none;
}

.c8:hover:not([disabled]),
.c8:focus-visible:not([disabled]) {
  color: #404040;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-family: 'Poppins',sans-serif;
  position: relative;
}

.c4 {
  color: #151515;
  font-family: 'Poppins',sans-serif;
  font-weight: 600;
  font-size: 16px;
  line-height: 20px;
  margin-bottom: 8px;
}

.c6 {
  color: #151515;
  background-color: #F9F9F9;
  width: 100%;
  max-width: 100%;
  min-width: 100%;
  font-size: 18px;
  font-family: 'Poppins',sans-serif;
  font-weight: 600;
  line-height: 22px;
  border-radius: 4px;
  border: 1px solid #151515;
  padding: 8px 16px;
  box-sizing: border-box;
  height: 48px;
}

.c6:disabled {
  background: #EEEEEE;
  border: 1px solid #A4A4A4;
}

.c6:invalid {
  border: 2px solid #BC2828;
}

.c6::-webkit-input-placeholder {
  color: #A4A4A4;
  font-size: 18px;
  font-weight: 400;
}

.c6::-moz-placeholder {
  color: #A4A4A4;
  font-size: 18px;
  font-weight: 400;
}

.c6:-ms-input-placeholder {
  color: #A4A4A4;
  font-size: 18px;
  font-weight: 400;
}

.c6::placeholder {
  color: #A4A4A4;
  font-size: 18px;
  font-weight: 400;
}

.c6:focus {
  outline: 1px solid #151515;
}

.c5 {
  position: relative;
}

.c0 {
  color: #151515;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 12px;
  color: #151515;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  width: 100%;
  gap: 12px;
  color: #151515;
}

.c7 {
  color: #151515;
  background: transparent;
  border: none;
  cursor: pointer;
  height: 48px;
  right: 0;
  top: 32px;
  position: relative;
  margin-left: -16px;
  -webkit-transform: scale(0.8);
  -ms-transform: scale(0.8);
  transform: scale(0.8);
}

<div>
    <p
      class="c0"
    >
      Custom title
    </p>
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for=""
          >
            Label
          </label>
          <div
            class="c5"
          >
            <input
              aria-invalid="false"
              aria-label="Label"
              class="c6"
              id=""
              name=""
              value=""
            />
          </div>
        </div>
        <div
          class="c3"
        >
          <label
            class="c4"
            for=""
          >
            Value
          </label>
          <div
            class="c5"
          >
            <input
              aria-invalid="false"
              aria-label="Value"
              class="c6"
              id=""
              name=""
              value=""
            />
          </div>
        </div>
        <button
          aria-label="Remove property"
          class="c7"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6L6 18"
              stroke="#151515"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M6 6L18 18"
              stroke="#151515"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </div>
    <button
      aria-label="Add Row"
      class="c8"
    >
      <span
        class="c9"
      >
        <img
          alt="Add Row icon"
          src="[object Object]"
        />
        Add Row
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`<Properties /> rendering should render properties and matches snapshot 1`] = `
<DocumentFragment>
  .c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex: unset;
  -ms-flex: unset;
  flex: unset;
  line-height: 1;
  gap: 8px;
}

.c8 {
  text-align: center;
  border: none;
  outline: none;
  overflow: hidden;
  cursor: pointer;
  pointer-events: auto;
  -webkit-transition: background-color 0.25s ease-out;
  transition: background-color 0.25s ease-out;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  min-height: 40px;
  font-weight: 600;
  padding: 0 28px;
  font-size: 16px;
  border-radius: 4px;
  background-color: transparent;
  padding: 2px 1px 2px 1px;
  min-height: 16px;
  border-radius: 0px;
  color: #151515;
}

.c8:disabled {
  opacity: 20%;
  cursor: default;
  pointer-events: none;
}

.c8:hover:not([disabled]),
.c8:focus-visible:not([disabled]) {
  color: #404040;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-family: 'Poppins',sans-serif;
  position: relative;
}

.c4 {
  color: #151515;
  font-family: 'Poppins',sans-serif;
  font-weight: 600;
  font-size: 16px;
  line-height: 20px;
  margin-bottom: 8px;
}

.c6 {
  color: #151515;
  background-color: #F9F9F9;
  width: 100%;
  max-width: 100%;
  min-width: 100%;
  font-size: 18px;
  font-family: 'Poppins',sans-serif;
  font-weight: 600;
  line-height: 22px;
  border-radius: 4px;
  border: 1px solid #151515;
  padding: 8px 16px;
  box-sizing: border-box;
  height: 48px;
}

.c6:disabled {
  background: #EEEEEE;
  border: 1px solid #A4A4A4;
}

.c6:invalid {
  border: 2px solid #BC2828;
}

.c6::-webkit-input-placeholder {
  color: #A4A4A4;
  font-size: 18px;
  font-weight: 400;
}

.c6::-moz-placeholder {
  color: #A4A4A4;
  font-size: 18px;
  font-weight: 400;
}

.c6:-ms-input-placeholder {
  color: #A4A4A4;
  font-size: 18px;
  font-weight: 400;
}

.c6::placeholder {
  color: #A4A4A4;
  font-size: 18px;
  font-weight: 400;
}

.c6:focus {
  outline: 1px solid #151515;
}

.c5 {
  position: relative;
}

.c0 {
  color: #151515;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 12px;
  color: #151515;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  width: 100%;
  gap: 12px;
  color: #151515;
}

.c7 {
  color: #151515;
  background: transparent;
  border: none;
  cursor: pointer;
  height: 48px;
  right: 0;
  top: 32px;
  position: relative;
  margin-left: -16px;
  -webkit-transform: scale(0.8);
  -ms-transform: scale(0.8);
  transform: scale(0.8);
}

<div>
    <p
      class="c0"
    >
      Properties
    </p>
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
        >
          <label
            class="c4"
            for=""
          >
            Label
          </label>
          <div
            class="c5"
          >
            <input
              aria-invalid="false"
              aria-label="Label"
              class="c6"
              id=""
              name=""
              value=""
            />
          </div>
        </div>
        <div
          class="c3"
        >
          <label
            class="c4"
            for=""
          >
            Value
          </label>
          <div
            class="c5"
          >
            <input
              aria-invalid="false"
              aria-label="Value"
              class="c6"
              id=""
              name=""
              value=""
            />
          </div>
        </div>
        <button
          aria-label="Remove property"
          class="c7"
        >
          <svg
            fill="none"
            height="24"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6L6 18"
              stroke="#151515"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
            <path
              d="M6 6L18 18"
              stroke="#151515"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
    </div>
    <button
      aria-label="Add Row"
      class="c8"
    >
      <span
        class="c9"
      >
        <img
          alt="Add Row icon"
          src="[object Object]"
        />
        Add Row
      </span>
    </button>
  </div>
</DocumentFragment>
`;
